samples/nrf9160/nrf_cloud_mqtt_multi_service/src/connection.c

```diff
@@ -0,0 +1,740 @@
+	if (ret) {
+		LOG_ERR("Transmission of enqueued device message failed, nrf_cloud_send "
+			"gave error: %d. The message will be re-enqueued and tried again "
+			"later.", ret);
+
+		/* Re-enqueue the message for later retry (creating a new copy of it in heap). */
+		send_device_message(msg);
+
+		/* Increment the failure counter. */
+		send_failure_count += 1;
+
+		/* If we have failed too many times in a row, there is likely a bigger problem,
+		 * and we should reset our connection to nRF Cloud, and wait for a few seconds.
+		 */
+		if (send_failure_count > CONFIG_MAX_CONSECUTIVE_SEND_FAILURES) {
+			/* Disconnect. */
+			disconnect_cloud();
+
+			/* Wait for a few seconds before trying again. */
+			k_sleep(K_SECONDS(CONFIG_CONSECUTIVE_SEND_FAILURE_COOLDOWN_SECONDS));
+		}
+	} else {
+		LOG_DBG("Enqueued device message consumed successfully");
+
+		/* Reset the failure counter, since we succeeded. */
+		send_failure_count = 0;
+	}
+
+	/* Clean up. */
+	k_free(msg);
+
+	return ret;
+}
+
+int send_device_message(const char *const msg)
+{
+	/* Copy message data onto the heap (will be freed by the message consumer when done). */
+
+	char *msg_buf = k_malloc(strlen(msg) + 1);
+
+	if (!msg_buf) {
+		LOG_ERR("Could not alloc memory for new device message");
+		return -ENOMEM;
+	}
+	strcpy(msg_buf, msg);
+	LOG_DBG("Enqueued message: %s", log_strdup(msg_buf));
+
+	/* Attempt to append data onto message queue. */
+	if (k_msgq_put(&device_message_queue, &msg_buf, K_NO_WAIT)) {
+		LOG_ERR("Device message rejected, outgoing message queue is full");
+		k_free(msg_buf);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+int send_device_message_cJSON(cJSON *msg_obj)
+{
+	int ret = 0;
+	char *msg = NULL;
+
+	if (!msg_obj) {
+		LOG_ERR("Cannot send NULL device message object");
+		return -EINVAL;
+	}
+
+	/* Convert message object to a string. */
+	msg = cJSON_PrintUnformatted(msg_obj);
+	if (msg == NULL) {
+		LOG_ERR("Failed to convert cJSON device message object to string");
+		return -ENOMEM;
+	}
+
+	/* Send the string as a device message. */
+	ret = send_device_message(msg);
+
+	/* Clean up */
+	k_free(msg);
+	return ret;
+}
+
+/**
+ * @brief Reset connection to nRF Cloud, and reset connection status event state.
+ * For internal use only. Externally, disconnect_cloud() may be used instead.
+ */
+static void reset_cloud(void)
+{
+	int err;
+	bool badreset = false;
+
+	/* Wait for a few seconds to help residual events settle. */
+	LOG_INF("Resetting nRF Cloud transport");
+	k_sleep(K_SECONDS(20));
+
+	/* Disconnect from nRF Cloud and uninit the cloud library. */
+	err = nrf_cloud_uninit();
+
+	/* nrf_cloud_uninit returns -EBUSY if reset is blocked by a FOTA job. */
+	if (err == -EBUSY) {
+		LOG_ERR("Could not reset nRF Cloud transport due to ongoing FOTA job. "
+			"Continuing without resetting");
+	} else if (err) {
+		LOG_ERR("Could not reset nRF Cloud transport, error %d. "
+			"Continuing without resetting", err);
+		badreset = true;
```

Comment:
this variable isn't used.

Reply:
🤦removed, thank you for catching this.

I really **_do_** read my own code. Don't know how I miss stuff like this.